### PR TITLE
Enabled the efficient filtering support for Faiss Engine

### PR DESCRIPTION
**Pull request by navneet1v**
### Description
This is an initial PR for enabling the pre-filtering support for Faiss Engine.

#### Changes include:
1. Upgrading the Faiss to commit id: https://github.com/facebookresearch/faiss/commit/3219e3d12e6fc36dfdfe17d4cf238ef70bf89568
2. Doing an optimization to select when to use IDArraySelector and IDBitMapSelector.
3. Don't run Faiss search if no docIds are returned from filters.

## Next Steps:
1. Optimize the JNI interfaces to have single query interface.
2. Write Unit tests for both java and c++ code.
3. Write ITs for filtering.
5. Do perf testing, compare with lucene and share the results on the github issue.
6. Figure out better way to select IDSelector Map.
7. Implement Exact search optimization when filterIds < k

### Issues Resolved
#903 

###Testing
Tested the code using the example mentioned here: https://opensearch.org/docs/latest/search-plugins/knn/filter-search-knn/#using-a-lucene-k-nn-filter for lucene

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java [hunk 1]
```diff
@@ -191,6 +213,60 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
...
+        final int cost = acceptDocs.cardinality();
+        log.info("Number of docs valid for filter is = Cost for filtered k-nn is : {}", cost);
+        return acceptDocs;
+    }
+
+    private BitSet createBitSet(final DocIdSetIterator filteredDocIdsIterator, final Bits liveDocs, int maxDoc) throws IOException {
+        if (liveDocs == null && filteredDocIdsIterator instanceof BitSetIterator) {
+            // If we already have a BitSet and no deletions, reuse the BitSet
+            return ((BitSetIterator) filteredDocIdsIterator).getBitSet();
+        } else {
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/query/KNNWeight.java:237]: nit: is else condition necessary?

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java [hunk 2]
```diff
@@ -191,6 +213,60 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
...
+                protected boolean match(int doc) {
+                    return liveDocs == null || liveDocs.get(doc);
+                }
+            };
+            return BitSet.of(filterIterator, maxDoc);
+        }
+    }
+
+    private int[] getFilterIdsArray(final LeafReaderContext context) throws IOException {
+        int[] filteredIds = new int[0];
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/query/KNNWeight.java:250]: Can we flip if statement on this line to avoid long nested if?

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java [hunk 3]
```diff
@@ -191,6 +213,60 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
...
+            final BitSet filteredDocsBitSet = getFilteredDocsBitSet(context, this.filterWeight);
+            filteredIds = new int[filteredDocsBitSet.cardinality()];
+            int filteredIdsIndex = 0;
+            int docId = 0;
+            while (true) {
+                docId = filteredDocsBitSet.nextSetBit(docId);
+                if (docId == DocIdSetIterator.NO_MORE_DOCS || docId + 1 == DocIdSetIterator.NO_MORE_DOCS) {
+                    break;
+                }
+                log.debug("Docs in filtered docs id set is : {}", docId);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/query/KNNWeight.java:261]: What does this mean?

File: jni/include/org_opensearch_knn_jni_FaissService.h [hunk 4]
```diff
@@ -1,14 +1,3 @@
-/*
- * SPDX-License-Identifier: Apache-2.0
- *
- * The OpenSearch Contributors require contributions made to
- * this file be licensed under the Apache-2.0 license or a
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply jni/include/org_opensearch_knn_jni_FaissService.h:5]: sure. bad copy paste.

File: src/main/java/org/opensearch/knn/jni/FaissService.java [hunk 5]
```diff
@@ -83,6 +83,8 @@ public static native void createIndexFromTemplate(
      */
     public static native KNNQueryResult[] queryIndex(long indexPointer, float[] queryVector, int k);
 
+    public static native KNNQueryResult[] queryIndex_WithFilter(long indexPointer, float[] queryVector, int k, int[] filterIds);
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/jni/FaissService.java:86]: sure

File: jni/src/faiss_wrapper.cpp [hunk 6]
```diff
@@ -181,12 +188,73 @@ jlong knn_jni::faiss_wrapper::LoadIndex(knn_jni::JNIUtilInterface * jniUtil, JNI
 
 jobjectArray knn_jni::faiss_wrapper::QueryIndex(knn_jni::JNIUtilInterface * jniUtil, JNIEnv * env, jlong indexPointerJ,
                                                 jfloatArray queryVectorJ, jint kJ) {
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply jni/src/faiss_wrapper.cpp:190]: sure

File: jni/src/faiss_wrapper.cpp [hunk 7]
```diff
@@ -181,12 +188,73 @@ jlong knn_jni::faiss_wrapper::LoadIndex(knn_jni::JNIUtilInterface * jniUtil, JNI
...
+ *
+ * TODO: We need to ideally decide when we can take another hit of K iterations in latency. Some facts:
+ * an OpenSearch Index can have max segment size as 5GB which, which on a vector with dimension of 128 boils down to
+ * 7.5M vectors.
+ * Ref: https://opensearch.org/docs/latest/search-plugins/knn/knn-index/#hnsw-memory-estimation
+ * M = 16
+ * Dimension = 128
+ * (1.1 * ( 4 * 128 + 8 * 16) * 7000000)/(1000*1000*1000) ~ 4.9GB
+ * Ids are sequential in a Segment which means for IDSelectorBitmap total size if the max ID has value of 7M will be
+ * 7000000/(8*1000) = 875KBs in worst case.
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply jni/src/faiss_wrapper.cpp:211]: what do you mean by add offset to prevent excess storage?

File: jni/src/faiss_wrapper.cpp [hunk 8]
```diff
@@ -181,12 +188,73 @@ jlong knn_jni::faiss_wrapper::LoadIndex(knn_jni::JNIUtilInterface * jniUtil, JNI
...
+ *
+ * in 875KB how many ids can be represented as an array of 64-bit longs : 109,375 ids
+ * So iterating on 110k ids for 1 single pass is also time consuming
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply jni/src/faiss_wrapper.cpp:214]: I think, for this PR I will keep this open. We need to do benchmarks to conclude that. I will reword this to say what we are doing.

File: src/main/java/org/opensearch/knn/index/query/KNNScorer.java [hunk 9]
```diff
@@ -56,4 +56,37 @@ public float score() {
...
+            private final DocIdSetIterator docIdsIter = DocIdSetIterator.empty();
+
+            @Override
+            public DocIdSetIterator iterator() {
+                return docIdsIter;
+            }
+
+            @Override
+            public float getMaxScore(int upTo) throws IOException {
+                return Float.MIN_VALUE;
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/KNNScorer.java:77]: I want to return min, and in k-nn scores can be negative too, so I thought lets go with a Float MIN value.

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java [hunk 10]
```diff
@@ -191,6 +213,60 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
...
+        final Scorer scorer = filterWeight.scorer(ctx);
+        if (scorer == null) {
+            return new FixedBitSet(0);
+        }
+
+        final BitSet acceptDocs = createBitSet(scorer.iterator(), liveDocs, maxDoc);
+        // TODO: Based on this cost shift to exact search, because even in ANN search you have to calculate the
+        // distance for K vectors. This can avoid calls to native layer and save some latency.
+        final int cost = acceptDocs.cardinality();
+        log.info("Number of docs valid for filter is = Cost for filtered k-nn is : {}", cost);
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:229]: Yes this can be debug. I will change this.

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java (hunk 1 above)
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:237]: yes we can remove else.

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java (hunk 2 above)
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:250]: Yes sure,

File: jni/src/faiss_wrapper.cpp [hunk 11]
```diff
@@ -181,12 +188,73 @@ jlong knn_jni::faiss_wrapper::LoadIndex(knn_jni::JNIUtilInterface * jniUtil, JNI
...
+ * https://github.com/facebookresearch/faiss/wiki/Setting-search-parameters-for-one-query#idselectorarray-idselectorbatch-and-idselectorbitmap
+ *
+ * class	       storage	lookup     construction(Opensearch + Faiss)
+ * IDSelectorArray	O(k)	O(k)          O(2k)
+ * IDSelectorBatch	O(k)	O(1)          O(2k)
+ * IDSelectorBitmap	O(n/8)	O(1)          O(k) -> n is the max value of id in the index
+ *
+ * TODO: We need to ideally decide when we can take another hit of K iterations in latency. Some facts:
+ * an OpenSearch Index can have max segment size as 5GB which, which on a vector with dimension of 128 boils down to
+ * 7.5M vectors.
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply jni/src/faiss_wrapper.cpp:205]: fixed it. Typo from my side

File: src/main/java/org/opensearch/knn/index/query/KNNQueryFactory.java [hunk 12]
```diff
@@ -59,27 +59,42 @@ public static Query create(CreateQueryRequest createQueryRequest) {
...
 
         if (KNNEngine.getEnginesThatCreateCustomSegmentFiles().contains(createQueryRequest.getKnnEngine())) {
             log.debug(String.format("Creating custom k-NN query for index: %s \"\", field: %s \"\", k: %d", indexName, fieldName, k));
             return new KNNQuery(fieldName, vector, k, indexName);
         }
 
+        if (filterQuery != null) {
+            return new KnnFloatVectorQuery(fieldName, vector, k, filterQuery);
+        }
+        log.debug(String.format("Creating Lucene k-NN query for index: %s \"\", field: %s \"\", k: %d", indexName, fieldName, k));
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/KNNQueryFactory.java:75]: it is coming from here: https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/KNNQueryFactory.java#L81

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java [hunk 13]
```diff
@@ -191,6 +213,60 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
         return new KNNScorer(this, docIdSetIter, scores, boost);
     }
 
+    private BitSet getFilteredDocsBitSet(LeafReaderContext ctx, Weight filterWeight) throws IOException {
+        final Bits liveDocs = ctx.reader().getLiveDocs();
+        final int maxDoc = ctx.reader().maxDoc();
+
+        final Scorer scorer = filterWeight.scorer(ctx);
+        if (scorer == null) {
+            return new FixedBitSet(0);
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:222]: this is more like a defensive check, to see if scorer is null or not for a query. if it is null you return an empty bitset.

File: src/main/java/org/opensearch/knn/index/query/KNNScorer.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/KNNScorer.java:77]: Okay i will put 0.

File: src/main/java/org/opensearch/knn/jni/JNIService.java [hunk 14]
```diff
@@ -94,16 +94,21 @@ public static long loadIndex(String indexPath, Map<String, Object> parameters, S
...
+     * @param filteredIds  array of ints on which should be used for search.
      * @return KNNQueryResult array of k neighbors
      */
-    public static KNNQueryResult[] queryIndex(long indexPointer, float[] queryVector, int k, String engineName) {
+    public static KNNQueryResult[] queryIndex(long indexPointer, float[] queryVector, int k, String engineName, int[] filteredIds) {
         if (KNNEngine.NMSLIB.getName().equals(engineName)) {
             return NmslibService.queryIndex(indexPointer, queryVector, k);
         }
 
+        if (filteredIds != null && filteredIds.length != 0 && KNNEngine.FAISS.getName().equals(engineName)) {
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/jni/JNIService.java:108]: We already do that. Its just that it been done at other file level. Over here this check is added as a safety. We can remove the filterIds length == 0 check, as it is already taken care

File: src/main/java/org/opensearch/knn/index/query/KNNQueryFactory.java [hunk 15]
```diff
@@ -59,27 +59,42 @@ public static Query create(CreateQueryRequest createQueryRequest) {
         final String fieldName = createQueryRequest.getFieldName();
         final int k = createQueryRequest.getK();
         final float[] vector = createQueryRequest.getVector();
+        final Query filterQuery = getFilterQuery(createQueryRequest);
+        if (filterQuery != null && KNNEngine.getEnginesThatCreateCustomSegmentFiles().contains(createQueryRequest.getKnnEngine())) {
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/KNNQueryFactory.java:63]: I will see how i can simplify the checks

File: src/main/java/org/opensearch/knn/jni/JNIService.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/jni/JNIService.java:108]: You can check KNNWeight.java class. We dont load the files or do anything in case the filterids length is 0.

File: jni/src/faiss_wrapper.cpp [hunk 16]
```diff
@@ -195,14 +271,49 @@ jobjectArray knn_jni::faiss_wrapper::QueryIndex(knn_jni::JNIUtilInterface * jniU
...
+        int *filteredIdsArray = jniUtil->GetIntArrayElements(env, filterIdsJ, nullptr);
+        int filterIdsLength = env->GetArrayLength(filterIdsJ);
+        std::unique_ptr<faiss::IDSelector> idSelector;
+        std::string idSelectorType = getIdSelectorType(filteredIdsArray, filterIdsLength);
+        // start with empty vectors for 2 different types of empty Selectors. We need define them here to avoid copying of data
+        // during the returns. We could have used pass by reference, but we choose pointers. Returning reference to local
+        // vector is also an option which can be efficient than copying during returns but it requires upto date C++ compilers.
+        // To avoid all those confusions, its better to work with pointers here. Ref: https://cplusplus.com/forum/general/56177/
+        std::vector<faiss::idx_t> convertedIds;
+        std::vector<uint8_t> bitmap;
```
Comment (jmazanec15) [inline jni/src/faiss_wrapper.cpp:287]: Why do these need to be defined outside of the if statement?

File: jni/src/faiss_wrapper.cpp [hunk 17]
```diff
@@ -195,14 +271,49 @@ jobjectArray knn_jni::faiss_wrapper::QueryIndex(knn_jni::JNIUtilInterface * jniU
...
+            convertFilterIdsToFaissIdType(filteredIdsArray, filterIdsLength, convertedIds.data());
+            idSelector.reset(new faiss::IDSelectorBatch(convertedIds.size(), convertedIds.data()));
+        } else {
+            int maxIdValue = filteredIdsArray[filterIdsLength - 1];
+            const int bitsetArraySize = (maxIdValue >> 3) + 1;
+            bitmap.resize(bitsetArraySize, 0);
+            buildFilterIdsBitMap(filteredIdsArray, filterIdsLength, bitmap.data());
+            idSelector.reset(new faiss::IDSelectorBitmap(filterIdsLength, bitmap.data()));
+        }
+        std::unique_ptr<faiss::SearchParameters> hnswParams = buildSearchParams(indexReader, idSelector.get());
```
Comment (jmazanec15) [inline jni/src/faiss_wrapper.cpp:300]: nit: Can we call searchParams instead of hnswParams?

File: jni/src/faiss_wrapper.cpp [hunk 18]
```diff
@@ -179,14 +193,76 @@ jlong knn_jni::faiss_wrapper::LoadIndex(knn_jni::JNIUtilInterface * jniUtil, JNI
...
+ * So iterating on 117k ids for 1 single pass is also time consuming. So, we are currently concluding to consider only size
+ * as factor. We need to improve on this.
+ *
+ * TODO: Best way is to implement a SparseBitSet in C++. This can be done by extending the IDSelector Interface of Faiss.
+ *
+ * @param filterIds
+ * @param filterIdsLength
+ * @return std::string
+ */
+std::string getIdSelectorType(const int* filterIds, int filterIdsLength) {
```
Comment (jmazanec15) [inline jni/src/faiss_wrapper.cpp:225]: nit: can we move these private methods to below public methods?

File: jni/src/faiss_wrapper.cpp [hunk 19]
```diff
@@ -179,14 +193,76 @@ jlong knn_jni::faiss_wrapper::LoadIndex(knn_jni::JNIUtilInterface * jniUtil, JNI
...
+    int maxIdValue = filterIds[filterIdsLength - 1];
+    if(filterIdsLength * sizeof(faiss::idx_t) * 8 <= maxIdValue ) {
+        return "array";
```
Comment (jmazanec15) [inline jni/src/faiss_wrapper.cpp:228]: nit: use enum for IdSelectorType instead of string?

**Review by jmazanec15 — APPROVED** [review]
Overall, looks good! Added a few comments that can be addressed in future revisions.

File: jni/src/faiss_wrapper.cpp (hunk 16 above)
Comment (navneet1v) [reply jni/src/faiss_wrapper.cpp:287]: the problem is if we define inside of if statement, during memory crunch these objects memory is getting freed. Hence we need to define them outside so that till we do the search these variables are in scope.

File: jni/src/faiss_wrapper.cpp (hunk 17 above)
Comment (navneet1v) [reply jni/src/faiss_wrapper.cpp:300]: We can do that. it got missed. I will fix it.

File: jni/src/faiss_wrapper.cpp (hunk 18 above)
Comment (navneet1v) [reply jni/src/faiss_wrapper.cpp:225]: yes sure. let me do that.

File: jni/src/faiss_wrapper.cpp (hunk 19 above)
Comment (navneet1v) [reply jni/src/faiss_wrapper.cpp:228]: I wasn't sure if C++ supports enum. Let me see if it does and change it.

File: jni/src/faiss_wrapper.cpp (hunk 16 above)
Comment (navneet1v) [reply jni/src/faiss_wrapper.cpp:287]: I also added comments on top of it why I have done that.

**Comment by navneet1v** [comment]
> Overall, looks good! Added a few comments that can be addressed in future revisions.

resolved the comments and raised new revision

**Review by martin-gaievski — APPROVED** [review]
Overall it looks good, please try to address my comments for Java code in coming PRs
